File: CS/Form1.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Text;
using System.Windows.Forms;
using DevExpress.XtraTreeList.Nodes.Operations;
using DevExpress.XtraTreeList.Nodes;
using DevExpress.XtraTreeList.Data;
using DevExpress.XtraEditors;

namespace WindowsApplication1 {
    public partial class Form1 : XtraForm {
        public Form1() {
            InitializeComponent();
        }
        protected override void OnLoad(EventArgs e) {
            base.OnLoad(e);
            DataTable dt = new DataTable();
            dt.Columns.Add("ID", typeof(int));
            dt.Columns.Add("ParentID", typeof(int));
            dt.Columns.Add("ProductID", typeof(int));
            dt.Columns.Add("CountryID", typeof(string));

            dt.Rows.Add(new object[] { 1, DBNull.Value, 0, "us" });
            dt.Rows.Add(new object[] { 2, 1, 1, "us" });
            dt.Rows.Add(new object[] { 3, 1, 1, "uk" });
            dt.Rows.Add(new object[] { 4, 1, 2, "us" });
            dt.Rows.Add(new object[] { 5, 1, 2, "uk" });
            dt.Rows.Add(new object[] { 6, 1, 3, "ie" });
            dt.Rows.Add(new object[] { 7, DBNull.Value, 1, "uk" });
            dt.Rows.Add(new object[] { 8, 7, 4, "us" });
            dt.Rows.Add(new object[] { 9, 7, 4, "uk" });
            dt.Rows.Add(new object[] { 10, 7, 5, "us" });
            dt.Rows.Add(new object[] { 11, 7, 5, "uk" });
            dt.Rows.Add(new object[] { 12, 7, 5, "ie" });

            treeList1.KeyFieldName = "ID";
            treeList1.ParentFieldName = "ParentID";
            treeList1.DataSource = dt.DefaultView;
        }
        private void simpleButton1_Click(object sender, EventArgs e) {
            int productID    = Convert.ToInt32(spinEdit1.EditValue);
            string countryID = textEdit1.Text;
            //Solution1
            //int ID = 10;
            //treeList1.FocusedNode = treeList1.FindNodeByKeyID(ID);

            //Solution2
            TreeListOperationFindNodeByProductAndCountryValues findNode = new TreeListOperationFindNodeByProductAndCountryValues(productID, countryID);
            treeList1.NodesIterator.DoOperation(findNode);
            treeList1.FocusedNode = findNode.Node;
        }
    }
    public class TreeListOperationFindNodeByProductAndCountryValues : TreeListOperation {
        public const string ProductIDColumnName = "ProductID";
        public const string CountryIDColumnName = "CountryID";
        private TreeListNode nodeCore;
        private object productIDCore;
        private object countryIDCore;
        private bool isNullCore;
        public TreeListOperationFindNodeByProductAndCountryValues(object productID, object countryID) {
            this.productIDCore = productID;
            this.countryIDCore = countryID;
            this.nodeCore = null;
            this.isNullCore = TreeListData.IsNull(productIDCore) || TreeListData.IsNull(countryIDCore);
        }
        public override void Execute(TreeListNode node) {
            if (IsLookedFor(node.GetValue(ProductIDColumnName), node.GetValue(CountryIDColumnName)))
                this.nodeCore = node;
        }
        bool IsLookedFor(object productID, object countryID) {
            if (IsNull) return (productIDCore == productID && countryIDCore == countryID);
            return productIDCore.Equals(productID) && countryIDCore.Equals(countryID);
        }
        protected bool IsNull { get { return isNullCore; } }
        public override bool CanContinueIteration(TreeListNode node) { return Node == null; }
        public TreeListNode Node { get { return nodeCore; } }
    }
}

File: VB/Form1.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.Data
Imports System.Drawing
Imports System.Text
Imports System.Windows.Forms
Imports DevExpress.XtraTreeList.Nodes.Operations
Imports DevExpress.XtraTreeList.Nodes
Imports DevExpress.XtraTreeList.Data
Imports DevExpress.XtraEditors

Namespace WindowsApplication1
	Partial Public Class Form1
		Inherits XtraForm
		Public Sub New()
			InitializeComponent()
		End Sub
		Protected Overrides Sub OnLoad(ByVal e As EventArgs)
			MyBase.OnLoad(e)
			Dim dt As New DataTable()
			dt.Columns.Add("ID", GetType(Integer))
			dt.Columns.Add("ParentID", GetType(Integer))
			dt.Columns.Add("ProductID", GetType(Integer))
			dt.Columns.Add("CountryID", GetType(String))

			dt.Rows.Add(New Object() { 1, DBNull.Value, 0, "us" })
			dt.Rows.Add(New Object() { 2, 1, 1, "us" })
			dt.Rows.Add(New Object() { 3, 1, 1, "uk" })
			dt.Rows.Add(New Object() { 4, 1, 2, "us" })
			dt.Rows.Add(New Object() { 5, 1, 2, "uk" })
			dt.Rows.Add(New Object() { 6, 1, 3, "ie" })
			dt.Rows.Add(New Object() { 7, DBNull.Value, 1, "uk" })
			dt.Rows.Add(New Object() { 8, 7, 4, "us" })
			dt.Rows.Add(New Object() { 9, 7, 4, "uk" })
			dt.Rows.Add(New Object() { 10, 7, 5, "us" })
			dt.Rows.Add(New Object() { 11, 7, 5, "uk" })
			dt.Rows.Add(New Object() { 12, 7, 5, "ie" })

			treeList1.KeyFieldName = "ID"
			treeList1.ParentFieldName = "ParentID"
			treeList1.DataSource = dt.DefaultView
		End Sub
		Private Sub simpleButton1_Click(ByVal sender As Object, ByVal e As EventArgs) Handles simpleButton1.Click
			Dim productID As Integer = Convert.ToInt32(spinEdit1.EditValue)
			Dim countryID As String = textEdit1.Text
			'Solution1
			'int ID = 10;
			'treeList1.FocusedNode = treeList1.FindNodeByKeyID(ID);

			'Solution2
			Dim findNode As New TreeListOperationFindNodeByProductAndCountryValues(productID, countryID)
			treeList1.NodesIterator.DoOperation(findNode)
			treeList1.FocusedNode = findNode.Node
		End Sub
	End Class
	Public Class TreeListOperationFindNodeByProductAndCountryValues
		Inherits TreeListOperation
		Public Const ProductIDColumnName As String = "ProductID"
		Public Const CountryIDColumnName As String = "CountryID"
		Private nodeCore As TreeListNode
		Private productIDCore As Object
		Private countryIDCore As Object
		Private isNullCore As Boolean
		Public Sub New(ByVal productID As Object, ByVal countryID As Object)
			Me.productIDCore = productID
			Me.countryIDCore = countryID
			Me.nodeCore = Nothing
			Me.isNullCore = TreeListData.IsNull(productIDCore) OrElse TreeListData.IsNull(countryIDCore)
		End Sub
		Public Overrides Sub Execute(ByVal node As TreeListNode)
			If IsLookedFor(node.GetValue(ProductIDColumnName), node.GetValue(CountryIDColumnName)) Then
				Me.nodeCore = node
			End If
		End Sub
		Private Function IsLookedFor(ByVal productID As Object, ByVal countryID As Object) As Boolean
			If IsNull Then
				Return (productIDCore Is productID AndAlso countryIDCore Is countryID)
			End If
			Return productIDCore.Equals(productID) AndAlso countryIDCore.Equals(countryID)
		End Function
		Protected ReadOnly Property IsNull() As Boolean
			Get
				Return isNullCore
			End Get
		End Property
		Public Overrides Function CanContinueIteration(ByVal node As TreeListNode) As Boolean
			Return Me.Node Is Nothing
		End Function
		Public ReadOnly Property Node() As TreeListNode
			Get
				Return nodeCore
			End Get
		End Property
	End Class
End Namespace